homeassistant/components/coinbase/__init__.py

```diff
@@ -1,110 +1,128 @@
-    account_currencies = config[DOMAIN].get(CONF_ACCOUNT_CURRENCIES)
-    exchange_currencies = config[DOMAIN][CONF_EXCHANGE_CURRENCIES]
-
-    hass.data[DATA_COINBASE] = coinbase_data = CoinbaseData(api_key, api_secret)
-
-    if not hasattr(coinbase_data, "accounts"):
-        return False
-    for account in coinbase_data.accounts:
-        if account_currencies is None or account.currency in account_currencies:
-            load_platform(hass, "sensor", DOMAIN, {"account": account}, config)
-    for currency in exchange_currencies:
-        if currency not in coinbase_data.exchange_rates.rates:
-            _LOGGER.warning("Currency %s not found", currency)
-            continue
-        native = coinbase_data.exchange_rates.currency
-        load_platform(
-            hass,
-            "sensor",
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the Coinbase component."""
+    if DOMAIN not in config:
+        return True
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
             DOMAIN,
-            {"native_currency": native, "exchange_currency": currency},
-            config,
+            context={"source": SOURCE_IMPORT},
+            data=config[DOMAIN],
         )
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Coinbase from a config entry."""
+
+    client = await hass.async_add_executor_job(
+        Client,
+        entry.data[CONF_API_KEY],
+        entry.data[CONF_API_TOKEN],
+    )
+
+    hass.data.setdefault(DOMAIN, {})
+
+    hass.data[DOMAIN][entry.entry_id] = await hass.async_add_executor_job(
+        CoinbaseData, client
+    )
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
 
     return True
 
 
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    return unload_ok
+
+
+def get_accounts(client):
+    """Handle paginated accounts."""
+    response = client.get_accounts()
+    accounts = response[API_ACCOUNTS_DATA]
+    next_starting_after = response.pagination.next_starting_after
+
+    while next_starting_after:
+        response = client.get_accounts(starting_after=next_starting_after)
+        accounts += response[API_ACCOUNTS_DATA]
+        next_starting_after = response.pagination.next_starting_after
+
+    return accounts
+
+
 class CoinbaseData:
     """Get the latest data and update the states."""
 
-    def __init__(self, api_key, api_secret):
+    def __init__(self, client):
         """Init the coinbase data object."""
 
-        self.client = Client(api_key, api_secret)
-        self.update()
+        self.client = client
+        self.accounts = get_accounts(self.client)
+        self.exchange_rates = self.client.get_exchange_rates()
```

Comment:
We don't want side effects in init method. Please move those out and pass the result to the data class when instantiating it or move the side effects to a separate setup method.

Reply:
Is it ok to call that setup method from the init method? I originally moved these here rather than using the previous call to `self.update` to step around the throttling which was causing values not to be available for the length of the throttle interval at setup/boot

Reply:
Actually, I just tested using `self.update()` in place of:
```
        self.accounts = get_accounts(self.client)
        self.exchange_rates = self.client.get_exchange_rates()
```
and it seems fine actually. I presume it's ok to revert to that?

Reply:
Never mind, had a coffee and I got what you mean. Sorry